fields/comparison.go

```diff
@@ -0,0 +1,118 @@
+	CompareValuesForUpdate(field Field, newValue, oldValue *string) FieldsComparator
+	CompareValueAndFunction(field Field, newValue *string, oldValueFunc func(Field) []string) FieldsComparator
+	CompareValueAndFunctionForUpdate(field Field, newValue *string, oldValueFunc func(Field) []string) FieldsComparator
+	CompareOptionalAndFunction(field Field, newValue csjson.OptionalString, oldValueFunc func(Field) []string) FieldsComparator
+
+	IsAnyFieldUpdated(fields ...Field) bool
+	AreAllFieldsUpdated(fields ...Field) bool
+	UpdatedFields() []string
+}
+
+type fieldsComparator struct {
+	updatedFields map[Field]bool
+	caseSensitive bool
+}
+
+func NewFieldsComparator() FieldsComparator {
+	return &fieldsComparator{updatedFields: make(map[Field]bool), caseSensitive: true}
+}
+
+func (fc *fieldsComparator) CaseSensitive(caseSensitive bool) FieldsComparator {
+	fc.caseSensitive = caseSensitive
+	return fc
+}
+
+func (fc *fieldsComparator) compareStrings(value1, value2 string) bool {
+	if fc.caseSensitive {
+		return value1 == value2
+	}
+	return strings.ToLower(value1) == strings.ToLower(value2)
+}
+
+func (fc *fieldsComparator) CompareValues(field Field, newValue, oldValue *string) FieldsComparator {
+	if (newValue == nil && oldValue != nil) ||
+		(newValue != nil && oldValue == nil) ||
+		(newValue != nil && oldValue != nil && !fc.compareStrings(*newValue, *oldValue)) {
+		fc.updatedFields[field] = true
+	}
+	return fc
+}
+
+func (fc *fieldsComparator) CompareValuesForUpdate(field Field, newValue, oldValue *string) FieldsComparator {
+	// When specifying ForUpdate, we suppose that no update is asked if second value is nil
+	if newValue == nil {
+		return fc
+	}
+	return fc.CompareValues(field, newValue, oldValue)
+}
+
+func (fc *fieldsComparator) CompareValueAndFunction(field Field, newValue *string, oldValueFunc func(Field) []string) FieldsComparator {
+	var oldValues = oldValueFunc(field)
+	var l = len(oldValues)
+	if l == 0 {
+		return fc.CompareValues(field, newValue, nil)
+	} else if l == 1 {
+		return fc.CompareValues(field, newValue, &oldValues[0])
+	}
+	fc.updatedFields[field] = true
+	return fc
+}
+
+func (fc *fieldsComparator) CompareValueAndFunctionForUpdate(field Field, newValue *string, oldValueFunc func(Field) []string) FieldsComparator {
+	if newValue == nil {
+		// Nothing to updated
+		return fc
+	}
+	return fc.CompareValueAndFunction(field, newValue, oldValueFunc)
+}
+
+func (fc *fieldsComparator) CompareOptionalAndFunction(field Field, newValue csjson.OptionalString, oldValueFunc func(Field) []string) FieldsComparator {
+	if !newValue.Defined {
+		// Nothing to updated
+		return fc
+	}
+	return fc.CompareValueAndFunction(field, newValue.Value, oldValueFunc)
+}
+
+// Return true if one of the given fields has been updated...
+// If no field is provided in parameters, just check if any field has been updated
+func (fc *fieldsComparator) IsAnyFieldUpdated(fields ...Field) bool {
```

Comment:
I would suggest to rename to IsAnyFieldsUpdated. Maybe a method IsFieldUpdated(field Field) could also be good.

Reply:
done

Reply:
We are checking if ONE field in the given list has been updated => IsAnyFieldUpdated (plural would be used if it was AreAllFieldsUpdated but here, it is singular)